[RFC] Add proper expression parser

ZyX-I opened the pull request:
Tasks for the parser:

- [ ] Be part of the regular expression evaluation process.
- [x] Be part of the highlighting routines for `<C-r>=`, etc.
- [ ] Used to provide smarter completion.
- [ ] Lambdas should hold AST and not `return {expr}` string.
- [x] External API: ASYNC function that yields expressions AST (should be possible to have function completely thread-safe and not only ASYNC if I get rid of things like vim_isIDc).
- [x] Small isolated program which does AST parsing, likely even without printing, for fuzzers, or in case I ever have time to use things like [KLEE](http://klee.github.io/).

Minimal tasks before completing which PR must not be merged: highlighting and API.

Constraints:

- Of course, it should not crash.
- Parser should make it possible to parse, *highlight and complete* any byte sequence which user may input. Highlighting and completion should be sensible enough for any practical input (e.g. I may input `<C-r>=(|)` with `|` being a cursor position: this is not a valid expression, but it what parser will once receive for highlighting), AST for invalid input should be just sensible enough to allow sensible highlighting and completion.
- Parser should yield good error messages.
- No recursion.

ZyX-I committed:
viml/expressions: Add lexer with some basic tests

ZyX-I committed:
viml/parser: Handle encoding conversions

ZyX-I committed:
viml/parser: Add helper functions for highlighting

ZyX-I committed:
unittests: Move some functions into helpers modules

ZyX-I committed:
viml/parser/expressions: Start creating expressions parser

Currently supported nodes:

- Register as it is one of the simplest value nodes (even numbers are
  not that simple with that dot handling).
- Plus, both unary and binary.
- Parenthesis, both nesting and calling.

Note regarding unit tests: it stores data for AST in highlighting in
strings in place of tables because luassert fails to do a good job at
representing big tables. Squashing a bunch of data into a single string
simply yields more readable result.

ZyX-I commented:
In the current state it is not going to pass any tests due to not handled values in `switch()`.

ZyX-I committed:
kvec: Add kv_drop() which is to be used like `(void)kv_pop(kvec)`

ZyX-I commented:
Some minimal language is finished: supporting unary/binary plus and calling/nesting parenthesis alongside with a single actual value (register) makes me think I can proceed with extending current parser code to accept the rest of the language and not inventing something else. On the way is, in any order

1. checking whether I got handling operators precedence right by adding multiplication (others may wait until 3. is sorted out)
2. comma operator (it is simpler to express multiargument call as `Call(func, Comma(Comma(arg1, arg2), arg3))`, then use comma code for list and dictionary literals with minimal modifications)
3. `{` handling as the greatest ambiguity in VimL expressions: it is used for lambdas, dictionaries and for curly braces names at once, in some positions exactly any one of the listed entities may appear
4. actually getting completion working: my idea is that if completion is requested parser function needs to be supplied with what is there up to the cursor, at EOC parser will set up completion by analyzing what is there in the AST stack and what was the previous token
5. colon operator (either `Ternary(cond, Colon())` or for dictionary, same reasoning as for commas).

Rest of parsing should be handled only after implementing the above.

(BTW, I think it may be a good and relatively easy to test lexer with KLEE. The only thing I completely do not like is that they do not support llvm-4.0 yet and llvm-3.x is not slotted and thus may not be installed alongside with 4.0, so KLEE could only either run in docker on my machine or I need to compile LLVM myself.)

ZyX-I committed:
klee: Start preparing for klee tests

First stage: something compiling without klee, but with a buch of dirty
hacks - done.
Second stage: something running under klee, able to emit useful results,
but still using dirty hacks - done.

Third stage: make CMake care about clang argumnets - not done, may be
omitted if proves to be too hard. Not that klee can be run on CI in any
case.

ZyX-I commented:
Managed to get KLEE run some tests, though it did not find anything useful yet. At least, I could be made sure that lexer always skips something (i.e. will not make parser fall into the infinite loop) and does not crash. Should be more useful for the parser (when it is finished), also turning code which compiles for KLEE into an executable suitable for fuzzing is much easier.

bfredl commented on src/nvim/CMakeLists.txt:
```diff
@@ -81,6 +81,8 @@ foreach(subdir
         event
         eval
         lua
+        viml
```
what is the logic behind `eval` / `viml` split? Is `viml` for "abstract" parsing/manipulation etc while `eval` is for actual evaluation/execution? Or is it just "old" vs "new" code? Is it important enough to warrant the split?

ZyX-I replied:
`eval` was intended for expression evaluation only. `viml` is for the whole VimL, including Ex commands. The “expression evaluation” parts like parsing and execution are going to move to viml/ for sure, the core problem is that after the whole VimL is ported to a new parser, the difference between expression evaluation and Ex commands evaluation should be gone (both will use a new VM, whatever it will actually be).

What is left of eval is

1. `typval_T` and related values manipulations. It is more logical to move that near the VM, to viml/ as well (just eval/typval\* to viml/typval\* when viml/ will actually contain some executor, not much hassle).
2. A big bunch of `f_…` functions. My idea is that most logical choice for that is treating them just like `ex_…` functions used for Ex commands: scattered around the codebase, near the domain these functions belong to. But this is too heavy refactoring so probably somebody will move them to a separate file (as a part of #5081) where that will live from now on. File may be moved later.
3. Utilities for things mentioned above: currently existing executing parser, functions, etc.

So it is just “old” vs “new” code, with “new” not using “eval” because it is not about only the expressions. I do not want to have expressions parsing in `eval/` because it is going to be fairly isolated from what already is there, but sharing some things with Ex commands parser (for which there is no preexisting directory in any case).

ZyX-I commented:
The KLEE finally finished, no errors in lexer found. For some reason I did not have time automatically reported (zsh has a feature which makes it automatically act as if command was run like `time …` if it took more then $REPORTTIME seconds to finish which I have set to 5, with some other exceptions which should not apply to my script), but first file generated by the script (`build/klee/a.bc`) states to be created Sep 12 01:21, the last one (`build/klee/out/info`) states Sep 12 16:04: almost 15h.

ZyX-I commented:
Total generated 53 873 “test cases”, 221 MiB according to `du -hs`. Generated directory is greatly compressable: attached `tar.xz` archive is only 1.3 MiB. Archive attached in case somebody is interested what KLEE generates, though it is not very interesting: lots of “test cases”, LLVM assembly used, three things looking like log files, some coverage data in text format (not very readable) and KLEE own profiling data (rather top level, KLEE was being used not for profiling KLEE after all).

[klee.tar.xz.not.zip](https://github.com/neovim/neovim/files/1297344/klee.tar.xz.not.zip)

justinmk commented:
@ZyX-I so you plan to expose the AST somehow, e.g. as a dictionary , to plugins? Or is it possible?

ZyX-I commented:
I plan to expose it, but only AST (not lexer). In the future - via two functions, one for parsing expressions exclusively, like `eval()` (it has some differences regarding newline handling), another for parsing Ex commands.

And only via API, relying on the VimL function generator for VimL. Plus parser may be exposed via ASYNC function.

ZyX-I committed:
viml/parser/expressions: Add support for figure braces (three kinds)

oni-link commented on src/nvim/viml/parser/expressions.c:
```diff
@@ -452,6 +464,8 @@ LexExprToken viml_pexpr_next_token(ParserState *const pstate, const bool peek)
 //
 // Used highlighting groups and assumed linkage:
 //
+// NVimInternalError -> highlight as fg:red/bg:red
```
Is this correct, fg and bg having the same color?

ZyX-I replied:
As an internal error it should be something which will make user report it. Given that I do not want to spawn disturbing errors “invisible”, outstanding character is a good alternative.

ZyX-I replied:
Note that this is also the only group which I did not bother to (plan to) link to some preexisting group.

ZyX-I commented:
Got working figure braces (dictionary, curly-braces-names in value position, lambdas, but not complex curly-braces-names setups like `foo{"bar"}`) and commas (minus multiargument calls: should be working but not tested yet); operator priority code looks like something working as well (minus “no associativity” part). Most problematic parts to code in parser should be coded now, though there are still challenges like highlighting of string literals, floating-point numbers, ternary operator or non-associative operators (and, of course, making that pass linter which will not be happy with a big function out there).

And no completion code yet, though I have an idea how to code it which needs only minor modifications to what is already there.

ZyX-I committed:
viml/parser/expressions: Make commas actually work when calling

ZyX-I committed:
viml/parser/expressions: Make commas actually work when calling

ZyX-I committed:
viml/parser/expressions: Make curly braces name actually work

ZyX-I committed:
unittests: Add a way to print trace on regular error

oni-link commented:
Error while building with `ninja`:
```
CMake Error at /home/oni-link/git/neovim/cmake/RunXgettext.cmake:13 (message):
  xgettext failed to run correctly: /usr/bin/xgettext: Non-ASCII string at
  ../viml/parser/expressions.c:1274.

                     Please specify the source encoding through --from-code.
```
Message contains `’`:
```
                ERROR_FROM_NODE_AND_MSG(
                    new_top_node,
                    _("E15: Don’t know what figure brace means: %.*s"));

```

ZyX-I committed:
viml/parser/expressions: Add a way to adjust lexer

It also adds support for kExprLexOr which for some reason was forgotten.

It was only made sure that KLEE test compiles in non-KLEE mode, not that
something works or that KLEE is able to run tests.

ZyX-I committed:
test/helpers: Add format_string and format_luav

First intended to provide %r functionality like in Python (and also support for 
%*.*s, but this was not checked), second adds nice table formatting for use in 
cases similar to screen:snapshot_util().

ZyX-I committed:
unittests: Add support for dumping “expected” state

Purpose is similar to that of `screen:snapshot_util()`, but in different domain.

ZyX-I committed:
viml/parser/expressions: Add a way to represent tokens from C code

ZyX-I committed:
viml/parser/expressions: Fix call inside nested parenthesis

It may have incorrectly tried to call everything because of essentially “value” 
nodes being treated as not such.

ZyX-I committed:
viml/parser/expressions: Fix determining invalid commas/colons

ZyX-I committed:
viml/parser/expressions: Add support for ternary operator

ZyX-I committed:
viml/parser/expressions: Make sure that arrows outside lambda throw

ZyX-I committed:
viml/parser/expressions: Add support for comparison operators

ZyX-I commented:
Got working lambdas, dictionaries, curly braces names, function calls and nesting parenthesis, ternary operator, comparisons, unary/binary plus, operator associativity in general.

What is left of specifically the parser

- Dot handling.
- String literals handling.
- Finishing rest of “not interesting” operators.
- Determining completion context.
- Making all that pass the linter (it will not be happy to see that huge parser function).
- KLEE check (going to be very time-consuming; currently rerunning it for the lexer).

ZyX-I committed:
viml/parser/expressions: Forbid dot or alpha characters after a float

This is basically what Vim already does, in addition to forbidding floats should
there be a concat immediately before it.

ZyX-I committed:
viml/parser/expressions: Add support for the dot operator and numbers

brcolow commented on src/nvim/viml/parser/expressions.c:
```diff
@@ -1332,11 +1345,32 @@ ExprAST viml_pexpr_parse(ParserState *const pstate, const int flags)
+    // simply can’t: whether "d" is a dictionary is not known at the parsing
+    // time.
+    //
+    // Here example will always contain a concat with "a:2" sucking colon,
```
sucking?

ZyX-I committed:
viml/parser/expressions: Make lexer parse numbers, support non-decimal

ZyX-I committed:
viml/parser/expressions: Error out on multiple colons in a row

ZyX-I committed:
viml/parser/expressions: Add support for subscript and list literals

ZyX-I committed:
mbyte: Lint some functions which are to be copied for symbolic tests

ZyX-I committed:
edit: Lint some functions which are to be copied for symbolic tests

ZyX-I committed:
keymap: Lint some functions to be copied for symbolic tests

ZyX-I committed:
viml/parser/expressions: Add support for string parsing

ZyX-I committed:
kvec: Add kv_Z which is like kv_A, but zero is the last value

ZyX-I commented:
String and concat handling are now present, as well as subscripts. Highlighting of strings is not as good as it might be (may use ast_stack to highlight regexes at the right of =~ or inside e.g. match() function, printf format strings or expression strings like in map()), but it is not of a priority.

This mostly leaves “not interesting” operators until I may start to integrate parser into other things, first in `<C-r>=`/etc highlighting.

ZyX-I committed:
viml/parser/expressions: Finish parser

Note: formatc.lua was unable to swallow some newer additions to ExprASTNodeType 
(specifically `kExprNodeOr = '|'` and probably something else), so all `= …` 
were dropped: in any case they only were there in order to not bother updating 
viml_pexpr_debug_print_ast_node and since it is now known all nodes which will 
be present it is not much of an issue.

ZyX-I commented:
Finished parser, but most recent commit is not tested at all, except for “it compiles (now without warnings on not handled switch values)” and “existing parser/lexer tests can be run”.

ZyX-I committed:
klee: Fix some errors made in …parser.c

oni-link commented on src/nvim/viml/parser/expressions.c:
```diff
@@ -0,0 +1,2764 @@
+  FUNC_ATTR_WARN_UNUSED_RESULT
+{
+  static char buf[sizeof(int) * 3 + 1];
+  if (20 <= ch && ch < 127) {
```
Printable characters start at `32 (0x20)`.

ZyX-I committed:
viml/parser/expressions: Create tests for latest additions

ZyX-I committed:
unittests: Test some edge cases

ZyX-I committed:
viml/parser/expressions,tests: Add AST freeing, with sanity checks

ZyX-I committed:
viml/parser/expressions: Update some comments and add another check

ZyX-I committed:
charset: Fix out-of-bounds array access

It is incorrect to *first* access ptr[2] and *then* check whether maxlen allows 
it.

ZyX-I committed:
viml/parser/expressions,klee: Fix some problems found by KLEE run

ZyX-I committed:
Merge branch 'master' into expression-parser

ZyX-I committed:
keymap: Readd figure braces disappeared when resolving conflicts

ZyX-I committed:
viml/parser/expressions: Define east_node_type_tab only when needed

ZyX-I committed:
charset: Avoid overflow in vim_str2nr

ZyX-I committed:
charset: Do not call strlen() from vim_str2nr

ZyX-I committed:
viml/parser/expressions: Highlight prefix separately from number

Should make accidental octals more visible.

ZyX-I committed:
klee: Include colors in test

ZyX-I committed:
klee: Update key_name_entry table

ZyX-I committed:
keymap: Remove incorrect cast

justinmk commented on src/nvim/keymap.c:
```diff
@@ -318,73 +316,73 @@ static struct mousetable {
+///
+/// E.g. 'S' for shift, 'C' for ctrl.
 int name_to_mod_mask(int c)
+  FUNC_ATTR_CONST FUNC_ATTR_WARN_UNUSED_RESULT
```
I thought "const" func attr cannot be used if globals (`mod_mask_table `) are referenced?

ZyX-I replied:
It is immutable global.

justinmk commented on src/nvim/keymap.c:
```diff
@@ -573,11 +571,7 @@ int find_special_key(const char_u **srcp, const size_t src_len, int *const modp,
-        } else {
-          l = 1;
-        }
+        l = utfc_ptr2len_len(bp + 1, (int)(end - bp) + 1);
```
I think we already agreed to keep the `mb_` names. `utf_` will be removed.

ZyX-I replied:
It is very bad idea, `mb_…` references *both* `utf` and `utfc` and they have different enough semantics. So I never use this variant.

justinmk commented on src/nvim/edit.c:
```diff
@@ -6066,27 +6066,30 @@ void free_last_insert(void)
-
-  len = (*mb_char2bytes)(c, temp);
-  for (i = 0; i < len; ++i) {
+  const int len = utf_char2bytes(c, temp);
```
`mb_`

ZyX-I commented:
Unless we also agree on removing `utfc` from `mb_` functions I am not going to ever use that. Can’t find where, but I remember a relevant bug. Also while seeking for that bug I found e9e1668ca: seemingly innocent commit slightly changes the function result: where function previously outright removed composing characters it now does not. Maybe it is that bug, in fact PR where this commit is present has a reference to vim/vim#1827 which describes exactly the problem, but commit message despite being my own does not mention any bugs.

ZyX-I committed:
test/unit/formatc: Fix parsing of most recent viml_parser_highlight

ZyX-I committed:
viml/parser/expressions: Fix memory leak when processing ternary

ZyX-I committed:
klee: Fix possible assertion error

No idea how it did not happen to hit me yet.

ZyX-I committed:
unittests: Free everything and check for memory leaks

Also improves error reporting.

ZyX-I committed:
klee: Add run.sh --help and run.sh -s

ZyX-I committed:
viml/parser/expressions: Remove unused flag

ZyX-I committed:
viml/parser/expressions: Add some casts

ZyX-I committed:
viml/parser/expressions: Fix some errors spotted by KLEE

Not all of them are fixed yet though.

oni-link commented on src/nvim/charset.c:
```diff
@@ -1655,88 +1658,66 @@ void vim_str2nr(const char_u *const start, int *const prep, int *const len,
             break;
           }
-          if (ptr[n] >= '0') {
+          if (ptr[i] >= '0') {
```
This condition is always true.

ZyX-I replied:
Wondering why it was there in the first place then and what was supposed to mean.

Vim has it as well though, so can ask there.

oni-link commented on src/nvim/charset.c:
```diff
@@ -1621,32 +1621,35 @@ bool vim_isblankline(char_u *lbuf)
   // Recognize hex, octal and bin.
-  if ((ptr[0] == '0') && (ptr[1] != '8') && (ptr[1] != '9')
-      && (maxlen == 0 || maxlen > 1)) {
+  if ((what & (STR2NR_HEX|STR2NR_OCT|STR2NR_BIN))
```
`STR2NR_ALL` could be used for `(STR2NR_HEX|STR2NR_OCT|STR2NR_BIN)`.

ZyX-I committed:
charset: Some more refactoring of vim_str2nr

ZyX-I committed:
viml/parser/expressions: Fix last error found by KLEE

ZyX-I committed:
unittests: Avoid alloc log checking errors when printing tests

ZyX-I committed:
klee: Update vim_str2nr in mock as well

ZyX-I committed:
ex_getln: Make use of new parser to color expressions

Retires g:Nvim_color_expr callback.

ZyX-I committed:
gen_declarations: Do not generate line numbers by default

ZyX-I commented:
I have added expressions coloring, though there is a question now: where do I define highlight groups, currently listed in expressions.c? So far I think that the best idea is to create a file filled with `hi def link` in `runtime/colors/nvim/parser.vim` and source that from `load_colorscheme()`.

justinmk commented:
@ZyX-I Is this not something one would expect to work without VIMRUNTIME?

ZyX-I committed:
api/vim: Add nvim_parse_expression function

ZyX-I commented:
@justinmk It is not going to work without syncolor.vim. I have put a thing into C code, but it appears that it only works properly after loading syncolor.vim.

ZyX-I committed:
syntax,viml/expressions/parser: Create defaults for expr highlighting

ZyX-I committed:
syntax: Adjust position and arguments of syn_init_cmdline_highlight

This way it works both after `nvim -u NORC` and after that and `colorscheme 
wombat256mod`. Removed the comment because I do not actually know why it works 
here with these arguments and not in previous position with previous arguments.

ZyX-I committed:
syntax,viml/parser/expressions: Add missing highlight groups

Also adjusts some names.

ZyX-I commented:
This is how it looks with `nvim -u NORC -i NONE<CR>:<C-r>=(2 + 2 * 2) * &test`:

![example highlighting](http://img-fotki.yandex.ru/get/892702/9151298.7/0_c7055_e4a5900c_orig.png)

(yes, parser does not know which options are valid so does not highlight &test as something invalid).

ZyX-I committed:
Merge branch 'master' into expression-parser

ZyX-I committed:
*: Fix linter errors

Big function in expressions.c may be refactored, if I ever catch the idea how to 
split it right.

ZyX-I committed:
ex_getln: Fix variable name which is wrong after the merge

ZyX-I committed:
tests: Fix testlint errors

oni-link commented on src/nvim/charset.c:
```diff
@@ -1621,131 +1621,94 @@ bool vim_isblankline(char_u *lbuf)
-        for (int n = 1; ascii_isdigit(ptr[n]); ++n) {
-          if (ptr[n] > '7') {
-            // can't be octal
+      if (what & STR2NR_OCT
```
If `STR2NR_FORCE` is also set, we later on always parse the string in base `8`. So checking for invalid digits would be unnecessary here.

oni-link commented on src/nvim/charset.c:
```diff
@@ -1621,131 +1621,94 @@ bool vim_isblankline(char_u *lbuf)
-          if (ptr[n] > '7') {
-            // can't be octal
+      if (what & STR2NR_OCT
+          && !STRING_ENDED(ptr + 1)
```
Could be removed, was already checked.

oni-link commented on src/nvim/api/vim.c:
```diff
@@ -886,6 +886,72 @@ Array nvim_call_atomic(uint64_t channel_id, Array calls, Error *err)
+///                    Use only "m" to parse like for "<C-r>=", only "E" to
+///                    parse like for ":echo", empty string for ":let".
+///
+/// @return AST: top-level dectionary holds keys
```
typo: dectionary -> dictionary

oni-link commented on src/nvim/generators/gen_declarations.lua:
```diff
@@ -164,9 +164,40 @@ local pattern = concat(
+
+    gendeclarations.lua definitions.c static.h non-static.h definitions.i
+
+Generates declarations for a C file defitions.c, putting declarations for
```
typo: defitions.c -> definitions.c

oni-link commented on src/nvim/generators/gen_declarations.lua:
```diff
@@ -164,9 +164,40 @@ local pattern = concat(
+Generates declarations for a C file defitions.c, putting declarations for
+static functions into static.h and declarations for non-static functions into
+non-static.h. File `definitions.i' should contain an already preprocessed
+version of defintions.c and it is the only one which is actually parsed,
```
typo: defintions.c -> definitions.c

oni-link commented on src/nvim/generators/gen_declarations.lua:
```diff
@@ -164,9 +164,40 @@ local pattern = concat(
+    NVIM_GEN_DECLARATIONS_LINE_NUMBERS:
+        If set to 1 then all generated declarations receive a comment with file
+        name and line number after the declaration. This may be useful for
+        debugging gen_declarations script, but not much beyound that with 
```
typo: beyound -> beyond

oni-link commented on src/nvim/edit.c:
```diff
@@ -6066,27 +6066,30 @@ void free_last_insert(void)
 char_u *add_char2buf(int c, char_u *s)
+  FUNC_ATTR_NONNULL_ALL FUNC_ATTR_WARN_UNUSED_RESULT
 {
   char_u temp[MB_MAXBYTES + 1];
```
`utf_char2bytes()` needs a buffer of at most size `6` (actually `4` if we follow the Unicode standard), not 
`MB_MAXBYTES +1 (=22)`.

oni-link commented on src/nvim/ex_getln.c:
```diff
@@ -2442,13 +2500,8 @@ static bool color_cmdline(CmdlineInfo *colored_ccline)
-    tl_ret = try_leave(&tstate, &err);
-    can_free_cb = true;
+    color_expr_cmdline(colored_ccline, ccline_colors);
+    can_free_cb = false;
```
Could be removed.

oni-link commented on src/nvim/keymap.c:
```diff
@@ -702,33 +696,39 @@ int find_special_key_in_table(int c)
+///                   with "t_" the next two characters are interpreted as
+///                   a termcap name.
+///
+/// @return Key code or 0 if ton found.
```
typo: ton -> not

ZyX-I committed:
charset: Refactor vim_str2nr

oni-link commented on src/nvim/charset.c:
```diff
@@ -1621,131 +1622,133 @@ bool vim_isblankline(char_u *lbuf)
-        }
+      goto vim_str2nr_bin;
+    }
+    // Detect octal number: zero followed by octal digits without '8' or '9'
```
The following digit could be for example 'A'. So it could be a non octal digit.

oni-link commented on src/nvim/charset.c:
```diff
@@ -1621,131 +1622,133 @@ bool vim_isblankline(char_u *lbuf)
+    if (!(what & STR2NR_OCT)) {
+      goto vim_str2nr_dec;
+    }
+    for (int i = 2; !STRING_ENDED(ptr + i) && ascii_isdigit(ptr[i]); i++) {
```
What about input like `0Ax`? This should not be recognized as octal number.

ZyX-I committed:
api/vim: Create part of nvim_parse_expression function

ZyX-I committed:
vim/api: Actually dump AST, fix some bugs in nvim_parse_expression

ZyX-I committed:
api/vim,functests: Add tests for nvim_parse_expression, fix found bugs

ZyX-I committed:
tests: Add missing test cases

ZyX-I committed:
api/vim: Add “len” dictionary key

This allows determining where parsing ended which may be needed for e.g. parsing
`:echo` with that API function.

ZyX-I committed:
viml/parser/expressions: Briefly document some differences

ZyX-I committed:
charset: Move vim_str2nr flags from vim.h to charset.h

ZyX-I committed:
unittests: Add tests for vim_str2nr

ZyX-I committed:
*: Fix some typos found by oni-link

ZyX-I committed:
Merge branch 'master' into expression-parser

ZyX-I committed:
*: Fix linter errors

ZyX-I committed:
unittests: Add a way to show some custom messages only when crashed

ZyX-I committed:
viml/parser/expressions: Replace lambda-specific WantedNode entries

This way code will be easier to adapt to handling (partially) non-expressions 
like :let lvalue part or :function definitions, and that would be needed in the 
future both for proper completion support and for the Ex commands parser.

ZyX-I committed:
viml/parser/expressions: Add support for parsing assignments

ZyX-I committed:
unittests: Add some more edge test cases

ZyX-I committed:
unittests,syntax: Check for sanity of highlight_init_cmdline

Also fixes some errors found.

ZyX-I committed:
unittests: Fix automatic test case generation

ZyX-I committed:
unittests,viml/parser/expressions: Start adding asgn parsing tests

ZyX-I commented:
Got KLEE run for a lexer again, no errors still (but it skipped some states though). Parser obviously does have errors, one of the commented tests crashes (and others are simply wrong). Though with that addition of parsing assignments PR came back from “almost ready to merge, need to fix a few FIXME’s” to “not finished” state, it should allow for more easy creation of both `:let`/`:for` parser and completion refactoring (latter probably not in this PR as I do not understand the relevant code enough above the point “it would need parsing `:let` to successfully complete”).

[klee-lexer.tar.xz.not.zip](https://github.com/neovim/neovim/files/1472724/klee-lexer.tar.xz.not.zip)

ZyX-I committed:
viml/parser/expressions,unittests: Do better testing, fix found issues

ZyX-I committed:
charset: Add missing include needed for vim_str2nr

ZyX-I committed:
functests: Fix ui/cmdline test

ZyX-I committed:
tests: Use single test file for unit and functional parser tests

ZyX-I committed:
*: Fix linter errors

ZyX-I committed:
ex_getln: Fix memory leak in color_expr_cmdline

ZyX-I committed:
tests: Stabilize float format and %e in format_luav and format_string

ZyX-I committed:
unittests: Avoid infinite cycle somewhere because of init failure

ZyX-I committed:
syntax: Fix duplicate group definitions

ZyX-I committed:
Merge branch 'master' into expression-parser

ZyX-I committed:
cmake: Fix api/version test failure

ZyX-I committed:
unittests: Populate ARGTYPES in child process only

ZyX-I committed:
doc: Update documentation

ZyX-I committed:
unittests: Run vim_str2nr tests with GC enabled

brcolow commented on runtime/doc/eval.txt:
```diff
@@ -10516,5 +10516,124 @@ This is not allowed when the textlock is active:
+
+Expressions entered by user in |i_CTRL-R_=|, |c_CTRL-\_e|, |quote=| are 
+colored by built-in expressions parser.  It uses highlight groups described in 
+the below table, which may be overriden by user colorschemes, all linked to 
```
"below table" => "table below" (can be both ways, but latter is a bit more Englishy)

brcolow commented on runtime/doc/eval.txt:
```diff
@@ -10516,5 +10516,124 @@ This is not allowed when the textlock is active:
+							*hl-NVimInvalid*
+In addition to highlighting groups prefixed with NVim described below there 
+are highlighting groups prefixed with NVimInvalid which have just the same 
+meaning, but used to indicate that relevant token contains an error or that 
```
"that relevant" => "that the relevant" (subjective)

ZyX-I committed:
Merge branch 'master' into expression-parser

ZyX-I committed:
doc: Some small fixes

ZyX-I commented:
Wondering whether it is safe to use xmalloc() in FUNC_API_ASYNC functions. And where is documentation for FUNC_API_…?

ZyX-I committed:
viml/parser/expressions: Make $ENV not depend on &isident

ZyX-I committed:
functests: Add some more tests

ZyX-I committed:
viml/parser/expressions: Make sure that listed nodes may be present

With the new test leaving `assert(false);` for any of the cases makes tests 
crash.

justinmk commented:
@ZyX-I because xmalloc calls `try_to_free_memory` / `preserve_exit` ? Good question. We could have a version that queues those things on the main loop, but the main loop may not be fully processed before `preserve_exit` is triggered elsewhere ...

> And where is documentation for FUNC_API_…?

Don't think there is any, unless @bfredl added some in one of his related PRs.

bfredl commented:
I should remember to do it in #7382

bfredl commented:
Also if xmalloc was somehow event unsafe I would have expected a lot of trouble already. It should be safe to always call on the main thread. NVIM_API_ASYNC is not as dangerous as it sounds, most things should be ok except for executing arbitrary API calls, vimL  and lua.

ZyX-I commented:
@justinmk Yes, because of that. After I got rid of &isident influence my parser function should not use anything which can’t be run in whatever thread you wish, except for allocator.

ZyX-I commented:
This PR should be ready if CI succeeds. Using parser for completion, expression evaluation, lambdas should go to another PR as this one is already big.

I think that next PR should better be switching from current `evalN` to some simple AST execution (not VM, at least yet), for some time hidden behind `set _vimlexecutor` similar to `set re` (though with underscore for it being temporary and not subject to backwards compatibility: I do not think anybody wants to maintain two VimL implementations simultaneously). I also do not think new variant may provide any benefits for anything, but lambdas - not as long as `:function` is a list of strings, and it will be a list of strings until there is a full parser.

ZyX-I commented:
Hm. apt-get failure on travis, something strange on AppVeyor:

```
[  ERROR   ] .\test\unit\viml\expressions\parser_tests.lua @ 4850: api nvim_parse_expression works with strings
C:/projects/neovim/test/functional\api\vim_spec.lua:880: Error while processing test ('"\\<C-u>"', ):
C:/projects/neovim/test/functional\api\vim_spec.lua:867: Expected objects to be the same.
Passed in:
(table) {
 *[ast] = {
   *[1] = 'DoubleQuotedString(val="<C-u>"):0:0:"\<C-u>"' } }
Expected:
(table) {
 *[ast] = {
   *[1] = 'DoubleQuotedString(val="\21"):0:0:"\<C-u>"' } }
stack traceback:
	C:/projects/neovim/test/functional\api\vim_spec.lua:880: in function 'check_parsing'
	.\test\unit\viml\expressions\parser_tests.lua:5620: in function <.\test\unit\viml\expressions\parser_tests.lua:4850>
```

What could convert `\21` to `<C-u>` (unit tests do not have this problem)? Same thing in QB, on my machine and, I think, on travis if it was working.

justinmk commented:
@ZyX-I did you enable unit tests on appveyor? Up to now they did not run on appveyor. 

Also it's weird to see unit-test path `.\test\unit\viml\expressions\parser_tests.lua ` and functional-test path `test/functional\api\vim_spec.lua` in that error message. Is this expected?

justinmk commented on runtime/doc/eval.txt:
```diff
@@ -10573,5 +10573,124 @@ This is not allowed when the textlock is active:
+the table below, which may be overriden by user colorschemes, all linked to 
+some other highlighting group.
+							*hl-NVimInvalid*
+In addition to highlighting groups prefixed with NVim described below there 
```
I don't like to nitpick about this, but here we are introducing yet another variant of "Nvim" spelling. Same with the highlight group names.

justinmk commented on src/nvim/version.c:
```diff
@@ -47,37 +47,39 @@ char *version_cflags = "Compilation: " NVIM_VERSION_CFLAGS;
 // clang-format off
 static const int included_patches[] = {
+  1230,
+  1229,
```
It's more important to include these in some commit message as vim-patch tokens:

```
vim-patch:8.0.1229
vim-patch:8.0.1230
```

Because soon version.c update will be automated, and it only looks at commit message tokens.

(Whoever merges this PR can just paste the above into the merge commit)

ZyX-I commented:
All parser tests are run two times: as unit via cffi and as functional test via API call (with a few bits of information missing as insignificant for API users, and also without explicit memory leak check and some tests impossible to perform via API). To avoid duplication I placed that in one file, near original location: there is no place for common *test* code, only for common utilities.AppVeyor runs only functional variant, but common code is in unit directory.21:24, 26 ноября 2017 г., "Justin M. Keyes" <notifications@github.com>:@ZyX-I did you enable unit tests on appveyor? Up to now they did not run on appveyor.Also it's weird to see unit-test path .\test\unit\viml\expressions\parser_tests.lua and functional-test path test/functional\api\vim_spec.lua in that error message. Is this expected?—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.

ZyX-I commented:
Checked log, it looks like the problem happens somewhere in C code:

```
2017/11/30 01:06:17 DEBUG 20824 log_client_msg:704: RPC <-ch 1: [request]  [0, 257, "nvim_parse_expression", [""\<C-u>"", "", true]]
2017/11/30 01:06:17 DEBUG 20824 log_server_msg:667: RPC ->ch 1: [response] [1, 257, nil, {"ast"=>{"type"=>"DoubleQuotedString", "start"=>[0, 0], "len"=>8, "svalue"=>"<C-u>"}, "len"=>8, "highlight"=>[[0, 0, 1, "NVimDoubleQuote"], [0, 1,3, "NVimDoubleQuotedUnknownEscape"], [0, 3, 7, "NVimDoubleQuotedBody"], [0, 7, 8, "NVimDoubleQuote"]]}]
```

(assuming logger does not do any escaping).

ZyX-I commented:
Well, it appears that &isident … is responsible for determining identifier length inside `<…>`. Since I set it to an empty string for testing only `-` could be present inside `<…>` then, and only because hyphen-minus is being tested for separately. I think that this is a bug because it is not mentioned in documentation and I see no good reasons for this option’s existence in any case.

ZyX-I committed:
keymap: Do not use vim_isIDc in keymap.c

Note: there are three changes to ascii_isident. Reverting first two (in 
find_special_key and first in get_special_key_code) normally fails the new test 
with empty &isident, but reverting the third does not. Hence adding `>` to 
&isident.

Ref vim/vim#2389.

ZyX-I committed:
Merge branch 'master' into expression-parser

ZyX-I committed:
unittests: Reduce memory used by vim_str2nr test

ZyX-I committed:
*: Replace all occurrences of NVim with Nvim

ZyX-I commented:
Got that after building libnvim-test on travis:

```
+asan_check /home/travis/build/neovim/neovim/build/log
+check_logs /home/travis/build/neovim/neovim/build/log '*san.*'
++find /home/travis/build/neovim/neovim/build/log -type f -name '*san.*'
+for log in '$(find "${1}" -type f -name "${2}")'
+sed -i /home/travis/build/neovim/neovim/build/log/ubsan.28775 -e '/Warning: noted but unhandled ioctl/d' -e '/could cause spurious value errors to appear/d' -e '/See README_MISSING_SYSCALL_OR_IOCTL for guidance/d'
+local err=
++find /home/travis/build/neovim/neovim/build/log -type f -name '*san.*' -size +0
+for log in '$(find "${1}" -type f -name "${2}" -size +0)'
+cat /home/travis/build/neovim/neovim/build/log/ubsan.28775
==28775==LeakSanitizer has encountered a fatal error.
==28775==HINT: For debugging, try setting environment variable LSAN_OPTIONS=verbosity=1:log_threads=1
==28775==HINT: LeakSanitizer does not work under ptrace (strace, gdb, etc)
```

What is that?

In any case, I am restarting travis under assumption that problem is temporary. Other CIs succeeded.

ZyX-I committed:
fix! set lsan options

ZyX-I commented:
With verboser logging:

```
+cat /home/travis/build/neovim/neovim/build/log/ubsan.28765
==28765==AddressSanitizer: failed to intercept '__isoc99_printf'
==28765==AddressSanitizer: failed to intercept '__isoc99_sprintf'
==28765==AddressSanitizer: failed to intercept '__isoc99_snprintf'
==28765==AddressSanitizer: failed to intercept '__isoc99_fprintf'
==28765==AddressSanitizer: failed to intercept '__isoc99_vprintf'
==28765==AddressSanitizer: failed to intercept '__isoc99_vsprintf'
==28765==AddressSanitizer: failed to intercept '__isoc99_vsnprintf'
==28765==AddressSanitizer: failed to intercept '__isoc99_vfprintf'
==28765==AddressSanitizer: failed to intercept '__cxa_throw'
==28765==AddressSanitizer: libc interceptors initialized
|| `[0x10007fff8000, 0x7fffffffffff]` || HighMem    ||
|| `[0x02008fff7000, 0x10007fff7fff]` || HighShadow ||
|| `[0x00008fff7000, 0x02008fff6fff]` || ShadowGap  ||
|| `[0x00007fff8000, 0x00008fff6fff]` || LowShadow  ||
|| `[0x000000000000, 0x00007fff7fff]` || LowMem     ||
MemToShadow(shadow): 0x00008fff7000 0x000091ff6dff 0x004091ff6e00 0x02008fff6fff
redzone=16
max_redzone=2048
quarantine_size_mb=256M
thread_local_quarantine_size_kb=1024K
malloc_context_size=30
SHADOW_SCALE: 3
SHADOW_GRANULARITY: 8
SHADOW_OFFSET: 0x7fff8000
==28765==Installed the sigaction for signal 11
==28765==Installed the sigaction for signal 7
==28765==Installed the sigaction for signal 8
==28765==T0: stack [0x7ffdf1066000,0x7ffdf1866000) size 0x800000; local=0x7ffdf1864260
==28765==LeakSanitizer: Dynamic linker not found. TLS will not be handled correctly.
==28765==AddressSanitizer Init done
==28766==Could not attach to thread 28765 (errno 1).
==28766==Failed suspending threads.
==28765==LeakSanitizer has encountered a fatal error.
==28765==HINT: For debugging, try setting environment variable LSAN_OPTIONS=verbosity=1:log_threads=1
==28765==HINT: LeakSanitizer does not work under ptrace (strace, gdb, etc)
```

ZyX-I committed:
Merge branch 'master' into expression-parser

Hoping that could fix the LSAN issue: no idea what it is talking about.

ZyX-I committed:
Revert "fix! set lsan options"

This reverts commit 62993323494d846190590606d37aff1136421671.

ZyX-I commented:
The problem somehow solved itself.

ZyX-I commented:
New build failure looks unrelated:

```
./test/functional/helpers.lua:268: 
retry() attempts: 2
./test/functional/ui/screen.lua:302: Row 2 did not match.
Expected:
  |{1: }                                                 |
  |*lost                                              |
  |gained                                            |
  |{4:~                                                 }|
  |{5:[No Name] [+]                                     }|
  |:                                                 |
  |{3:-- TERMINAL --}                                    |
Actual:
  |{1: }                                                 |
  |*gained                                            |
  |lost                                              |
  |gained                                            |
  |{5:[No Name] [+]                                     }|
  |:                                                 |
  |{3:-- TERMINAL --}                                    |

To print the expect() call that would assert the current screen state, use
screen:snapshot_util(). In case of non-deterministic failures, use
screen:redraw_debug() to show all intermediate screen states.  

stack traceback:
	./test/functional/helpers.lua:268: in function 'retry'
	...uild/neovim/neovim/test/functional/terminal/tui_spec.lua:300: in function <...uild/neovim/neovim/test/functional/terminal/tui_spec.lua:293>
```

ZyX-I committed:
tests: Fix linter errors

ZyX-I commented:
CI succeeded.

justinmk commented:
FYI the https://github.com/neovim/neovim/pull/7234#issuecomment-348785874 failure is my fault, I will make another change to the TUI startup and see if it continues; if it continues I will change the test to avoid spurious failure.

justinmk commented on runtime/doc/eval.txt:
```diff
@@ -10573,5 +10573,124 @@ This is not allowed when the textlock is active:
 	- etc.
 
+==============================================================================
+13. Command-line expressions coloring			*expr-coloring*
```
@ZyX-I Is there a reason to call it "coloring" instead of "highlighting"?

justinmk commented:
@ZyX-I what is your plan for command parsing or general VimL parsing? Will this be a new API function that accepts entire sources?

Is there a need for `nvim_parse_*` or would it make sense to have just one API function: `nvim_parse` .

oni-link commented on src/nvim/charset.c:
```diff
@@ -1611,141 +1611,145 @@ bool vim_isblankline(char_u *lbuf)
+      ptr++; \
+    } \
+  } while (0)
+  switch (pre) {
```
The `switch` could be replaced with
```c
vim_str2nr_bin: 
   PARSE_NUMBER(...);
   goto xxx;
vim_str2nr_oct:
   PARSE_NUMBER(...);
   goto xxx;
vim_str2nr_dec: 
   PARSE_NUMBER(...);
   goto xxx;
vim_str2nr_hex:
   PARSE_NUMBER(...);
xxx:
```

ZyX-I commented:
@justinmk There are use-cases for separate expression parser in the API other then making it easier to test some things (I mean, test as a plugin developer, Neovim developers are always going to have unit tests available), you may see a list in the documentation of the new function. For VimL parser I plan on exactly the API function accepting entire source, though internally it should be possible to save state and use it later to proceed (needed for alternative syntax highlighting: if we have a parser and highlighter in a C code, why limit highligter to command-line only?).
